Add nomination cog

ChrisLovering opened the pull request:
This cog listens for votes to be posted by the post in the nomination channel. When one is found, it creates a thread and pings admins/mods.

ChrisLovering committed:
Actually use loguru

ChrisLovering committed:
Add nomination cog

This cog listens for votes to be posted by the post in the nomination
channel. When one is found, it creates a thread and pings admins/mods.

ChrisLovering committed:
Add valid thread archive times enum

ChrisLovering committed:
Archive threads when the parent message is deleted

ChrisLovering committed:
Strip back intents to just what's needed

ChrisLovering committed:
Add more-itertools

ChrisLovering committed:
Add thread helper functions

ChrisLovering committed:
Refactor nomination cog to use helper functions

Bluenix2 commented on pyproject.toml:
```diff
@@ -16,6 +16,7 @@ statsd = "^=3.3.0"
 "discord.py" = {url = "https://github.com/Rapptz/discord.py/archive/master.zip"}
 loguru = "^0.5.3"
 pyyaml = "^=5.4.1"
+more-itertools = "^8.8.0"
```
I find this dependency rather useless, we use one function. Could easily make it ourselves.

ChrisLovering replied:
While I agree our current usage of it is minimal, however the intention is to eventually port this code over to the bot (When d.py 2.0 fully releases). Since bot has this package available, making use of it now to make more readable code is desirable.

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+    def __init__(self, bot: ThreadBot):
+        self.bot = bot
+        # Stores the name of the member recently posted nomination as,
+        # nomiantion messages can be split into multiple messages.
```
This comment doesn't really make sense?

Do you mean that this stores the most recent member that a nomination was posted about?

ChrisLovering replied:
Updated in https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1 think this is better now.

Leaving this open for feedback on new comment.

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+        self.bot = bot
+        # Stores the name of the member recently posted nomination as,
+        # nomiantion messages can be split into multiple messages.
+        self.nominated_member_name: int = None
```
This claims to be a name, but you type as an integer?

Additionally, you default this to None so it has to be `Optional[int]`.

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+
+        if not message.content.endswith(NOMINATION_ENDING_TEXT):
+            # Only create thread on final message, as a vote could be split into multiple messages.
+            return
```
Does this not create race conditions?

I don't know how you guys vote on nominations and how that system fully works, can there only ever be one nomination happening at a time?

ChrisLovering replied:
If a vote message is too long (due to very long nomination reasons from multiple people) the bot posts the vote over 2 or more messages. This logic is here so that we get the name of the nominated member from the first message, then wait for the last message to be detected, and make a thread on that.

I don't think we need to worry about 2 nominations coming in at the same time causing an issue here, as its *extremely* unlikely to get conditions that allow for that.

Bluenix2 replied:
Then I think for safety that before setting a new name that you have detected, check that it it is None and do `logger.error()` if not.

Bluenix2 replied:
Then I think for safety that before setting a new name that you have detected, check that it it is None and do `logger.error()` if not.

ChrisLovering replied:
Makes sense, added.

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+            # Only create thread on final message, as a vote could be split into multiple messages.
+            return
+        elif not self.nominated_member_name:
+            logger.error("Valid end message found, but no cached member name to create thread!")
```
Probably should return before creating a thread and causing issues?

ChrisLovering replied:
If we find a valid end message, I think we should still make the thread, the only issue will be that the thread title won't have the nominated member's name, which can be manually resolved

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+        if channel_id != constants.Channels.nomination_voting:
+            return  # Ignore messages deleted in other channels
+
+        if payload.cached_message and not payload.cached_message.flags.has_thread:
```
```suggestion
        elif payload.cached_message and not payload.cached_message.flags.has_thread:
```

ChrisLovering replied:
I'm not sure what this adds? Since the above if statement returns anyway, so if that's true this if will never be evaluated.

Bluenix2 replied:
Not much actually, usually I do like to use `elif`s in these cases to be clear that "Either this happens, or this".

I think we can leave it as is though yeah.

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+            return  # Ignore cached messages that are marked as not having threads
+
+        channel: discord.TextChannel = self.bot.get_channel(channel_id)
+        thread: discord.Thread = await get_thread_from_message_id(message_id, channel, self.bot.cached_messages)
```
When you override the typing like this, the if-statement below doesn't make sense.
```suggestion
        thread = await get_thread_from_message_id(message_id, channel, self.bot.cached_messages)
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+
+    If chunk_size is specified, chunked_find() will yield control to the event loop between each chunk.
+    """
+    for chunk in more_itertools.chunked(seq, chunk_size):
```
```py
def _chunk(seq: Iterable[T], size: int) -> Generator[Iterable[T]]:
    for i in range(0, len(seq), size):
        yield seq[i:i+size]
```
This should do the same without that dependency.

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+
+
+async def _check_first_message_referencing(thread: discord.Thread, message_id: int) -> Optional[discord.Thread]:
+    """Returns a thread if the thread_starter_message in `thread` references the message_id."""
```
I realize that if we're here, we can simply return a bool.
```suggestion
    """Return a bool indicating if the thread starter message references the given ``message_id`."""
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+    return None
+
+
+async def _check_first_message_referencing(thread: discord.Thread, message_id: int) -> Optional[discord.Thread]:
```
```suggestion
async def _check_first_message_referencing(thread: discord.Thread, message_id: int) -> bool:
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+    messages = await thread.history(limit=1, oldest_first=True).flatten()
+    if messages[0].reference.message_id == message_id:
+        return thread
+    return None
```
```suggestion
    return messages[0].reference.message_id == message_id
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+    channel: discord.TextChannel,
+    cached_messages: list[discord.Message]
+) -> Optional[discord.Thread]:
+    """Returns the thread linked to the given message id."""
```
```suggestion
    """Attempt to find the thread linked to the given message id."""
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+) -> Optional[discord.Thread]:
+    """Returns the thread linked to the given message id."""
+    def predicate(message: discord.Message) -> bool:
+        """Returns true if `message` is the thread_starter_message for the `message_id`."""
```
As a rule of thumb, I always stay away from "return**s**". The same way you don't use it in commits.

Instead of "If applied this commit will", think of it as "If called this function will".

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+            and isinstance(message.channel, discord.Thread)
+        )
+
+    # Try and find the thread start messag ein message cache.
```
```suggestion
    # Try and find the thread start message in message cache.
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+
+    # Try and find the thread start messag ein message cache.
+    # Use a chunked find as this many checks could block for too long.
+    if message := await chunked_find(predicate, cached_messages, chunk_size=200):
```
Could put this number in config, or constants?

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,71 @@
+        if found := await _check_first_message_referencing(thread, message_id):
+            logger.info("Thread found in fetched threads!")
+            return found
+    return None
```
I would like a blank like to kind of seperate this.
```suggestion
            return found

    return None
```

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class Nominations(commands.Cog):
+    """Cog for getting information about PyPi packages."""
```
This doesn't hold up, are you copying code Chris!!!

ChrisLovering replied:
lol, the ruse it up! I'm actually 3 hamsters in a trench coat copying code from everywhere!

ChrisLovering replied:
https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+class Nominations(commands.Cog):
+    """Cog for getting information about PyPi packages."""
+
+    def __init__(self, bot: ThreadBot):
```
```suggestion
    def __init__(self, bot: ThreadBot) -> None:
```

ChrisLovering replied:
https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+        # nomiantion messages can be split into multiple messages.
+        self.nominated_member_name: int = None
+
+        if constants.DEBUG_MODE:
+            self.archive_time = constants.ThreadArchiveTimes.DAY.value
+        else:
+            self.archive_time = constants.ThreadArchiveTimes.WEEK.value
```
Some suggestions, not sure what's best really..
```suggestion
        self.archive_time = constants.ThreadArchiveTimes.DAY.value if constants.DEBUG_MODE else constants.ThreadArchiveTimes.WEEK.value
```
```suggestion
        seconds = constants.ThreadArchiveTimes.DAY.value if constants.DEBUG_MODE else constants.ThreadArchiveTimes.WEEK.value
        self.archive_time = seconds
```
```suggestion
        archive_time = constants.ThreadArchiveTimes.DAY if constants.DEBUG_MODE else constants.ThreadArchiveTimes.WEEK
        self.archive_time = archive_time.value
```
I guess it may simply be best the way it's done

ChrisLovering replied:
I ended up with what I've got since I didn't want to deal with a very long conditional expression (which broke the line limit)

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+
+    @commands.Cog.listener()
+    async def on_message(self, message: discord.Message) -> None:
+        """Creates a thread whenever a vote is sent in the nominations voting channel."""
```
```suggestion
        """Create a thread on votes sent in the nominations voting channel."""
```

ChrisLovering replied:
https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+    async def on_message(self, message: discord.Message) -> None:
+        """Creates a thread whenever a vote is sent in the nominations voting channel."""
+        if message.channel.id != constants.Channels.nomination_voting:
+            return  # Ignore messages not in voting channel.
```
```suggestion
            return  # Ignore messages not in the voting channel
```

ChrisLovering replied:
https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+
+        if not message.content.endswith(NOMINATION_ENDING_TEXT):
+            # Only create thread on final message, as a vote could be split into multiple messages.
+            return
```
I would prefer it inline, so I guess you can short the comment into something like this
```suggestion
            return  # Only create threads on the final message; votes could be split into multiple
```

ChrisLovering replied:
https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+
+    @commands.Cog.listener()
+    async def on_raw_message_delete(self, payload: discord.RawMessageDeleteEvent) -> None:
+        """Archives a thread linked to a nomination vote when the nomination is archived."""
```
```suggestion
        """Archive threads linked to nomination votes when the vote is archived."""
```

ChrisLovering replied:
https://github.com/python-discord/thread-bot/pull/1/commits/6577ad7f19c2463c4c1201b6cfb761c9dcf273b1

Bluenix2 reviewed (changes requested):
You already know what is incoming haha

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+    re.MULTILINE
+)
+
+# When nominations are posted manually, the Discord message box converts +1/-1 to the thumbs emojis
```
Could we not word this better?

Do you mean that when this string is sent to Discord, it converts the unicode to `:thumbsup:` equivalent?

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+
+
+class Nominations(commands.Cog):
+    """Cog for creating an archiving nomination threads when votes are posted/archived."""
```
```suggestion
    """Cog for creating and archiving nomination threads when votes are posted/archived."""
```

Bluenix2 commented on bot/exts/nominations.py:
```diff
@@ -0,0 +1,81 @@
+            self.nominated_member_name = match.group(1)
+
+        if not message.content.endswith(NOMINATION_ENDING_TEXT):
+            return  # Only create thread on final message, as a vote could be split into multiple messages
```
This should fit the line length?
```suggestion
            return  # Votes could be split into multiple messages, create the thread on the final message
```

ChrisLovering committed:
Log an error if a new vote is found, but we still have a name cached

This can happen if two votes come in overlapping, while rare, worth keeping an eye for.

Co-authored-by: Bluenix <bluenixdev@gmail.com>

ChrisLovering committed:
Capitalise constant in constants.py

ChrisLovering committed:
Use a constant for chunk size

ChrisLovering committed:
Remove unnecessary passing of thread objects

We're calling this function in an if statement, which has access to the thread in it's scope. So just returning a bool is sufficient

Bluenix2 commented on bot/constants.py:
```diff
@@ -237,12 +238,23 @@ class Roles(metaclass=YAMLGetter):
-]
+)
+
+# Number for elements to chunk a sequence into when using bot.utils.helpers.chunked_find
```
```suggestion
# Amount of elements in each chunk of a sequence when using bot.utils.helpers.chunked_find
```

Bluenix2 commented on bot/utils/helpers.py:
```diff
@@ -0,0 +1,70 @@
+
+
+async def _check_first_message_referencing(thread: discord.Thread, message_id: int) -> bool:
+    """Check if the thread_starter_message of `thread` references the given ``message_id`."""
```
```suggestion
    """Check if the thread_starter_message of `thread` references the given `message_id`."""
```

Bluenix2 reviewed (dismissed):
Yeah this looks good now, thank you Chris
